### PR TITLE
check image readiness when recover

### LINKED ISSUE #303: Use containerd image readiness check function to properly recover image during restart.
Currently, during restart we read all image references from containerd image metadata store, and recover the image list.

However, a image reference in containerd image metadata store doesn't mean the image is ready for use, the image could be corrupted because of aborted image pulling.

We should use the newly added `Check` function https://github.com/containerd/containerd/pull/1567 to check content completeness and check top level snapshot existence.

### PR DESCRIPTION
**Pull request by yanxuean**
fix #303

Signed-off-by: yanxuean <yan.xuean@zte.com.cn>

### CODE REVIEW

File: pkg/server/restart.go [hunk 1]
```diff
@@ -101,6 +103,11 @@ func (c *criContainerdService) recover(ctx context.Context) error {
 		return fmt.Errorf("failed to load images: %v", err)
 	}
 	for _, image := range images {
+		if _, err := c.client.SnapshotService(c.config.ContainerdConfig.Snapshotter).Stat(ctx, image.ChainID); err != nil {
+			glog.Warningf("Failed to check top snapshot %+v: %v", image, err)
```
Comment (mikebrow) [inline pkg/server/restart.go]: top snapshot?

Comment (yanxuean) [reply pkg/server/restart.go]: Thanks. What do you suggest?

Comment (mikebrow) [reply pkg/server/restart.go]: "Failed to stat snapshot for image %+v: %v"
or 
"Failed to locate information for image %+v: %v"

Comment (yanxuean) [reply pkg/server/restart.go]: We want to check readiness of top snapshot.  We do it by calling Stat() function now.
Is the error message better to indicating our purpose?

Comment (mikebrow) [reply pkg/server/restart.go]: I understand your meaning and the issue description but most won't.  Top snapshot could have many meanings. Images have layers.. the image.chainid points to the top-level layer for an image snapshot and contains references to the rest of the layers. IMO it's better to just say what failed vs. using the issue description of what needed to be done here.  

For example:  "Failed to locate information for snapshot using chainid for image %+v: %v"
As another example: "Failed to stat the top-level snapshot for image %+v: %v"
I think those are more explanatory to the reader of the warnings than: "Failed to check top snapshot %+v: %v"

What do you think?

Comment (yanxuean) [reply pkg/server/restart.go]: You are right. You have a lot of experience,we need to learn from you.:)
Thank for your review and guidance.

Comment (yanxuean) [reply pkg/server/restart.go]: done

File: pkg/server/restart.go [hunk 2]
```diff
@@ -26,6 +26,8 @@ import (
 	"github.com/containerd/containerd"
 	"github.com/containerd/containerd/content"
 	"github.com/containerd/containerd/errdefs"
+	imagemanage "github.com/containerd/containerd/images"
```
Comment (Random-Liu) [inline pkg/server/restart.go]: This name is confusing. I prefer `containerdimages` if we really need another name for the package.

File: pkg/server/restart.go [hunk 3]
```diff
@@ -101,6 +103,12 @@ func (c *criContainerdService) recover(ctx context.Context) error {
 		return fmt.Errorf("failed to load images: %v", err)
 	}
 	for _, image := range images {
+		// Checking existence of top-level snapshot for each image being recovered.
```
Comment (Random-Liu) [inline pkg/server/restart.go]: Why not add this logic into `loadImages` function?

File: pkg/server/restart.go [hunk 4]
```diff
@@ -342,6 +350,10 @@ func loadImages(ctx context.Context, cImages []containerd.Image, provider conten
 		// imgs len must be > 0, or else the entry will not be created in
 		// previous loop.
 		i := imgs[0]
+		if ok, _, _, _, err := imagemanage.Check(ctx, provider, i.Target(), platforms.Default()); !ok || err != nil {
+			glog.Warningf("Failed to check image readiness for %q: %v", i.Name(), err)
```
Comment (Random-Liu) [inline pkg/server/restart.go]: Let's log different for `!ok` and `err != nil`.

Comment (Random-Liu) [reply pkg/server/restart.go]: Failed to check image content readiness for %q: %v

File: pkg/server/restart.go (hunk 3 above)
Comment (yanxuean) [reply pkg/server/restart.go]: Because we need call c.client.SnapshotService.

Comment (Random-Liu) [reply pkg/server/restart.go]: Just pass it in. :)

Comment (Random-Liu) [reply pkg/server/restart.go]: I feel like logically this should be part of `loadImages`, we load it only when it is ready.

Comment (yanxuean) [reply pkg/server/restart.go]: ok

File: pkg/server/restart.go (hunk 2 above)
Comment (yanxuean) [reply pkg/server/restart.go]: done

File: pkg/server/restart.go (hunk 4 above)
Comment (yanxuean) [reply pkg/server/restart.go]: done

**Comment by Random-Liu** [comment]
LGTM

**Pull request merged**